include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h

```diff
@@ -0,0 +1,1842 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2020 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+#ifndef dealii_mg_transfer_global_coarsening_templates_h
+#define dealii_mg_transfer_global_coarsening_templates_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/mpi_compute_index_owner_internal.h>
+#include <deal.II/base/mpi_consensus_algorithms.h>
+
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/dofs/dof_tools.h>
+
+#include <deal.II/fe/fe_tools.h>
+
+#include <deal.II/hp/dof_handler.h>
+
+#include <deal.II/matrix_free/evaluation_kernels.h>
+#include <deal.II/matrix_free/fe_evaluation.h>
+
+#include <deal.II/multigrid/mg_transfer_global_coarsening.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace
+{
+  /**
+   * Helper class to select the right templated implementation.
+   */
+  class CellTransfer
+  {
+  public:
+    CellTransfer(const unsigned int degree_fine,
+                 const unsigned int degree_coarse)
+      : n(100 * degree_fine + degree_coarse)
+    {}
+
+    template <typename Fu>
+    bool
+    run(Fu &fu)
+    {
+      return do_run_degree<1>(fu) || do_run_degree<2>(fu) ||
+             do_run_degree<3>(fu) || do_run_degree<4>(fu) ||
+             do_run_degree<5>(fu) || do_run_degree<6>(fu) ||
+             do_run_degree<7>(fu) || do_run_degree<8>(fu) ||
+             do_run_degree<9>(fu);
```

Comment:
I know that other cases are supposed to happen later; my question would be if that follow-up code would then go to a generic version with `-1` template argument (or `0` in the tensor product evaluators) so we do not bail out.

Reply:
I see your point. Let's leave it to a follow-up PR.